lib/github/merge/local.ex

```diff
@@ -0,0 +1,248 @@
+  @moduledoc """
+  Uses local Git to construct staging/trying branches instead of GitHub's
+  API.
+
+  Only used when local merges are enabled.
+  """
+
+  use GenServer
+
+  alias BorsNG.Database.Repo
+  alias BorsNG.Database.Batch
+  alias BorsNG.Database.Project
+  alias BorsNG.Database.LinkPatchBatch
+  alias BorsNG.GitHub
+  alias BorsNG.GitHub.Merge.Hooks
+  alias BorsNG.Worker.Batcher
+
+  @bors_author ["-c", "user.name='bors-ng'", "-c", "user.email='bors@app.bors.tech'"]
+
+  def start_link do
+    GenServer.start_link(__MODULE__, :ok, name: __MODULE__)
+  end
+
+  def init(_) do
+    {:ok, nil}
+  end
+
+  @doc """
+  GitHub allows GH Apps to authenticate over HTTPS using a username of
+  'x-access-token' and a password of the installation access token, which
+  happily gives us a way to authenticate CLI git commands.
+  """
+  def cli_creds(repo_conn) do
+    {:raw, token} = GitHub.get_raw_token!(repo_conn)
+    "x-access-token:#{token}"
+  end
+
+  @spec merge_batch!(Batch.t(), list(LinkPatchBatch.t()), %{commit: String.t(), tree: String.t()}) ::
+          %{commit: String.t(), tree: String.t()} | :conflict
+  def merge_batch!(batch, patch_links, base) do
+    GenServer.call(__MODULE__, {:merge_batch, batch, patch_links, base}, :infinity)
+  end
+
+  @spec squash_merge_batch!(Batch.t(), list(LinkPatchBatch.t()), map) ::
+          %{commit: String.t(), tree: String.t()}
+  def squash_merge_batch!(batch, patch_links, toml) do
+    GenServer.call(__MODULE__, {:squash_merge_batch, batch, patch_links, toml}, :infinity)
+  end
+
+  def handle_call(msg, _from, state) do
+    reply = do_handle_call(msg)
+    {:reply, reply, state}
+  end
+
+  defp do_handle_call({:merge_batch, batch, patch_links, base}) do
+    batch = batch |> Repo.preload([:project])
+    workdir = batch.project.name
+
+    repo_conn = Project.installation_connection(batch.project.repo_xref, Repo)
+
+    stmp = "#{batch.project.staging_branch}.tmp"
+
+    # Create the staging.tmp branch before we push to it
+    GitHub.synthesize_commit!(
+      repo_conn,
+      %{
+        branch: stmp,
+        tree: base.tree,
+        parents: [base.commit],
+        commit_message: "[ci skip][skip ci][skip netlify]",
+        committer: nil
+      }
+    )
+
+    # We make sure to clone the repository for each batch, rather than
+    # trying to reuse it, because hooks may change behavior in ways that could
+    # persist between batches in ways that Git can't detect, like modifying
+    # the .git/config.
+    File.rm_rf!(workdir)
+    init_project_repo!(batch.project, cli_creds(repo_conn))
+
+    git = fn args -> System.cmd("git", @bors_author ++ args, cd: workdir) end
+
+    git.(["fetch", "origin", batch.into_branch])
+    git.(["checkout", "origin/#{batch.into_branch}"])
+
+    Hooks.invoke_before_merge_hook!(workdir)
+
+    merge_status =
+      patch_links
+      |> Enum.reduce(:merged, fn link_patch_batch, status ->
+        case status do
+          :conflict ->
```

Comment:
do we need to handle any other statuses besides conflict and not-conflict? the api one also handles canceled

Reply:
weirdly enough, I can't find any way the equivalent API merge could receive a :canceled. none of the functions it calls return that. so that case seems to be a dead branch in the API merge?

for example, here's the signature of `merge_branch!`

```elixir
  @spec merge_branch!(tconn, %{
          from: bitstring,
          to: bitstring,
          commit_message: bitstring
        }) :: %{commit: binary, tree: binary} | :conflict
```

very strange.